### PR TITLE
search: Add search context support to experimental search input

### PR DESCRIPTION
**Pull request by fkling**
The current version of the search input prototype does not include a separate input element for the search context. Instead the search context is treated like any other filter, except we style it slightly differently (see video).

This diff does a couple of things:

- Adds support for value completion for the `context:` filter.
- Renders `context:` filters differently from other filters in the input.
- Refactors the whole suggestions implementation so that it will be easier to add new suggestion sources in the future.
- Introduces a custom version of the built-in placeholder extension that allows us to overwrite when/how the placeholder should be visible.

This diff also fixes an issue where we did not update the selected search context correctly when it wasn't present in a query.

I also had to update `@codemirror/view` and `@codemirror/state` to fix an issue with rendering decorations. This doesn't seem to have any effect on the current search input. Unfortunately it didn't fix all the issues (which I also point out in the video).

**NOTE**: Keep in mind that this is still a prototype and we are still verifying our approach. You might not agree with how search contexts work in this version, but that should not be the focus of the code review here.

https://user-images.githubusercontent.com/179026/212309355-1d8d1d63-5e11-4534-8cfc-4a337de82ce8.mp4

## Test plan

Manual testing.

## App preview:

- [Web](https://sg-web-fkling-experimental-search-context.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.06 kb) | 0.09% (+12.54 kb) 🔺 | 0.11% (+12.49 kb) 🔺 | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/fkling/experimental/search-context/compare-report.html;tab=live_object) for a full comparison between the commits 77cd409 and 03acd94 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by fkling** [comment]
Looks like prettier formatted pnpm-lock.yaml on my machine... should this file be formatted by prettier? @valerybugakov

**Comment by valerybugakov** [comment]
Good call: https://github.com/sourcegraph/sourcegraph/pull/46420

**Review by limitedmage — APPROVED** [review]
This is very nice.

Minor feedback (doesn't have to be fixed in this PR):

- There's an inconsistency with blurring: if you blur when the context is editable, the full context is still displayed. I think the blurred version of the input should be displayed the same no matter what the last thing edited was.
- The 'x' button to remove the context is not vertically-aligned
- I'm hoping that as a follow-up we can display contexts better. Having a floating `global` with no filter/prefix is odd/confusing; are we searching for the word "global" or are we searching in the "global" context? Rendering a display name and icon (eg. globe icon for "Global") might help with that.

File: client/web/src/search/home/SearchPage.tsx [hunk 1]
```diff
@@ -54,12 +54,21 @@ export const SearchPage: React.FunctionComponent<React.PropsWithChildren<SearchP
     const showCollaborators = window.context.allowSignup && homepageUserInvitation && props.isSourcegraphDotCom
     const { width } = useWindowSize()
     const shouldShowAddCodeHostWidget = useShouldShowAddCodeHostWidget(props.authenticatedUser)
+    const experimentalQueryInput = useExperimentalFeatures(features => features.searchQueryInput === 'experimental')
```
[the remark this answers is not in the document]
Comment (fkling) [reply client/web/src/search/home/SearchPage.tsx:57]: Yes, that's correct.

**Comment by fkling** [comment]
Thank you for the review @limitedmage !

> I'm hoping that as a follow-up we can display contexts better. Having a floating global with no filter/prefix is odd/confusing; are we searching for the word "global" or are we searching in the "global" context? Rendering a display name and icon (eg. globe icon for "Global") might help with that.

Definitely. I was implementation the original design, but I'm also not sure whether I really like hiding `context:`. Some stronger visual indicator is necessary IMO.

**Pull request merged**